File: gaTestSuiteMinimization.cpp

#include <iostream>
#include <fstream>
#include <string>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
#include <cstdlib>
#include <ctime>

#define N_INDIVIDUALS 100
#define MUTATION_RATE    0.01
#define CROSSOVER_PROBABILITY 0.9
#define INPUT_FILE "./input/matrix_94_3647_1.txt"
#define MAX_GENERATIONS 100

// Global Variables
int **currentGeneration;
int *generationFitness;
int totalOfFitnesses;
int numberEvaluations = 0;

int** reqMatrix;
int numberOfTests;
int numberOfReqs;

// Functions declarations
void allocateMemory(void);
int run(void);
void initializePopulation(void);
void evaluatePopulation(void);
void nextGen(void);
int selection(void);
int getBestSolution(void);
int getAverageSolution(void);
int eval(int*);
void mutate(int*);
int* crossover(int*, int*);
void displayMatrix(void);


// Allocate memory for our individuals
void allocateMemory(void){
    currentGeneration = new int*[N_INDIVIDUALS];

    generationFitness = new int[N_INDIVIDUALS];

    for(int index=0; index<N_INDIVIDUALS; ++index)
        currentGeneration[index] = new int[numberOfTests];
}


// Set random values to the individuals of the population
void initializePopulation(void){
    std::ifstream matrixFile(INPUT_FILE);
    std::string filename(INPUT_FILE);

    bool error = false;

    if(matrixFile.is_open()){

        numberOfReqs = stoi(filename.substr(filename.find("_", 0) + 1, filename.find("_", filename.find("_", 0) + 1))) + 1;
        numberOfTests = stoi(filename.substr(filename.find("_", filename.find("_", 0) + 1) + 1,
         filename.find("_", filename.find("_", filename.find("_", 0) + 1))));

        std::cout<<numberOfReqs<<":"<<numberOfTests<<std::endl;

        reqMatrix = new int*[numberOfTests];

        int i = 0;
        for (std::string line; std::getline(matrixFile, line); ){
            reqMatrix[i] = new int[numberOfReqs];

            for(int j = 0; j < numberOfReqs; ++j)
                reqMatrix[i][j] = line[j] - '0';

            ++i;
        }
    }
    else
        error = true;

    if(!error){

        // Allocate memory for our population
        allocateMemory();

        bool indexing[numberOfTests];
         
        for(int index=0; index<N_INDIVIDUALS; ++index){

            for(int i = 0; i < numberOfTests; ++i)
                indexing[i] = false;
            
            for(int gene=0; gene<numberOfTests; ++gene){
                int aux;

                do{
                    aux = rand() % numberOfTests;
                }while (indexing[aux]);

                indexing[aux] = true;

                currentGeneration[index][gene] = aux;
            }
        }
                
    }

}

// Evaluation of a individual
int eval(int* individual){
    int fitness = 0;

    bool satisfiedRequirements[numberOfReqs];
    for(int i = 0; i < numberOfReqs; ++i)
        satisfiedRequirements[i] = false;

    for(fitness = 0; fitness < numberOfTests; ++fitness){
        for(int j = 0; j < numberOfReqs; ++j)
            if(reqMatrix[individual[fitness]][j] == 1)
                satisfiedRequirements[j] = true;
        
        bool finish = true;
        for(int i = 0; i < numberOfReqs; ++i)
            if(!satisfiedRequirements[i]){
                finish = false;
                break;
            }

        if (finish)
            break;
    }

    numberEvaluations++;
    return fitness + 1;
}

// Returns the best fitness value
int getBestSolution(void) {
    float minFitness = numberOfTests;

    for (int i=0; i < N_INDIVIDUALS; ++i)
        if (generationFitness[i] < minFitness)
            minFitness = generationFitness[i];

    return minFitness;
}

// Returns the average fitness value
int getAverageSolution() {
  int avgFitness = 0;

  for (int i=0; i < N_INDIVIDUALS; ++i)
    avgFitness += generationFitness[i];

  avgFitness /= (double)N_INDIVIDUALS;

  return avgFitness;
}

// Evaluate all individuals
void evaluatePopulation(void){
  for(int i = 0; i < N_INDIVIDUALS; ++i)
    generationFitness[i] = eval(currentGeneration[i]);
}

// Binary Tournament Selection Operator
int selection(void){
    int i1 = rand() % (N_INDIVIDUALS);
    int i2;
    do{
        i2 = rand() % (N_INDIVIDUALS);
    }while(i1 == i2);

    if(generationFitness[i1] < generationFitness[i2])
        return i1;
    else
        return i2;
}

// PMX Crossover Operator
int * crossover(int* parent1, int* parent2){

    int point1 = rand() % numberOfTests;
    int point2;
    do{
        point2 = rand() % numberOfTests;
    }while(point1 == point2);

    int* offspring1 = new int[numberOfTests];
    int* offspring2 = new int[numberOfTests];

    std::vector<int> range;
    std::map<int, int> associations1;
    std::map<int, int> associations2;

    for(int i = point1 ; i < point2; i = (i + 1) % numberOfTests){
        range.push_back(i);

        associations1[parent1[i]] = parent2[i];
        associations2[parent2[i]] = parent1[i];

        offspring1[i] = parent2[i];
        offspring2[i] = parent1[i];
    }

    for(int i = 0; i < numberOfTests; ++i)
        if(std::find(range.begin(), range.end(), i) == range.end()){
            if(associations2.find(parent1[i]) != associations2.end()){
                int aux = parent1[i];
                do{
                    aux = associations2[aux];
                }while(associations2.find(aux) != associations2.end() && aux != parent1[i]);  
                offspring1[i] = aux;
            }
            else
                offspring1[i] = parent1[i];
            
            if(associations1.find(parent2[i]) != associations1.end()){
                int aux = parent2[i];
                do{
                    aux = associations1[aux];
                }while(associations1.find(aux) != associations1.end() && aux != parent2[i]);
                offspring2[i] = aux;
            }
            else
                offspring2[i] = parent2[i];

        } 

    int offFitness1 = eval(offspring1);
    int offFitness2 = eval(offspring2);

    if(offFitness1 < offFitness2){
        delete [] offspring2;
        return offspring1;
    }
    else{
        delete [] offspring1;
        return offspring2;
    }

}

// Mutation operator: random permutation of gene
void mutate(int* individual){
    for(int i=0; i<numberOfTests; ++i)
        if((rand()/(double)RAND_MAX) <= MUTATION_RATE) {
            int point1 = rand() % numberOfTests;
            int point2;
            do{
                point2 = rand() % numberOfTests;
            }while(point1 == point2);

            int aux = individual[point1];
            individual[point1] = individual[point2];
            individual[point2] = aux;
        }
}


// Evolve the population towards the next generation
void nextGen(void){
  int gene;
  int crossoverPoint;
  int mutateThisGene;

  for(int individual=0; individual<N_INDIVIDUALS; ++individual){

      // Selection
      int parent1 = selection();
      int parent2;
      do{
        parent2 = selection();
      }while(parent1 == parent2);

      // Crossover
      if( (rand()/(double)RAND_MAX) <= CROSSOVER_PROBABILITY){
        int* offspring = crossover(currentGeneration[parent1], currentGeneration[parent2]);

        // Mutation
        mutate(offspring);

        // Evaluation
        int fitnessOffspring = eval(offspring);

        // Recombination  
        if (fitnessOffspring <= generationFitness[individual]) { 
            delete [] currentGeneration[individual];
            currentGeneration[individual] = offspring;
            generationFitness[individual] = fitnessOffspring;
        }
        else
            delete [] offspring;
      }
    }

}

void displayMatrix(void){
    std::cout<<"#########################################"<<std::endl;
    for(int i = 0; i< N_INDIVIDUALS; ++i){
        for(int j = 0; j < numberOfTests; ++j)
            std::cout<<currentGeneration[i][j]<<",";
        std::cout<<std::endl;
    }
    std::cout<<"#########################################"<<std::endl;

    for(int i = 0; i< N_INDIVIDUALS; ++i){
        std::cout<<generationFitness[i];
        std::cout<<std::endl;
    }

    std::cout<<"#########################################"<<std::endl;

    for(int i = 0; i< numberOfTests; ++i){
        for(int j = 0; j < numberOfReqs; ++j)
            std::cout<<reqMatrix[i][j]<<",";
        std::cout<<std::endl;
    }

    std::cout<<"#########################################"<<std::endl;
}


// Performs the GA main loop: selection - recombination - mutation - replacement
int run(void){
    int generations = 0;

    initializePopulation();  

    // Iterate for a predefined maximum number of generations
    while(generations < MAX_GENERATIONS){
        evaluatePopulation();
        nextGen();
        //std::cout<<"\nGeneration "<<generations<<": Best solution: "<<getBestSolution()<<"; average fitness:"
            //<<getAverageSolution()<<std::endl;
        ++generations;

    }
    
    return generations;
}

int main(int argc, char *argv[]){
    int finalGeneration;
    time_t t, start, end;

    // Intialize random number generator 
    srand((unsigned) time(&t));

    time(& start);

    finalGeneration = run();

    time(& end);

    std::cout<<"\nThe best solution was found in generation "<<finalGeneration<<", with fitness: "
        <<getBestSolution()<<std::endl<<"Time: "<<difftime (end,start)<<" seconds"<<std::endl; 
}

File: gaTestSuiteMinimization.py
import time
import random

from tqdm import tqdm

N_INDIVIDUALS=100
MUTATION_RATE=0.01
CROSSOVER_PROBABILITY=0.9
INPUT_FILE="./input/matrix_94_3647_1.txt"
MAX_GENERATIONS=100

# Global Variables
currentGeneration = [] #list of list
generationFitness = []
totalOfFitnesses = 0
numberEvaluations = 0

reqMatrix = [] #list of list
numberOfTests = 0
numberOfReqs = 0

#functions

def run():

    initializePopulation()  

    for generations in tqdm(range(MAX_GENERATIONS)):
        evaluatePopulation()
        nextGen()
    
    return generations

def initializePopulation():
    error=False
    global numberOfReqs
    numberOfReqs = int(INPUT_FILE.split("_")[1])+1
    global numberOfTests
    numberOfTests = int(INPUT_FILE.split("_")[2])

    with open(INPUT_FILE, "r") as file:
        Lines = file.readlines()
    global reqMatrix
    reqMatrix =[]
    for line in Lines:
        reqMatrix.append(line)
    global currentGeneration
    for index in range(N_INDIVIDUALS):
        currentGeneration.append([])
        indexing = [False for i in range(numberOfTests)]

        for gene in range(numberOfTests):
            aux = random.randint(0,numberOfTests-1)
            while indexing[aux]:
                aux = random.randint(0,numberOfTests-1)
            indexing[aux] = True
            currentGeneration[index].append(aux)

def evaluatePopulation():
    global generationFitness
    for i in range(N_INDIVIDUALS):
        generationFitness.append(eval(currentGeneration[i]))
    
def nextGen():
    global currentGeneration
    global generationFitness
    for individual in range(N_INDIVIDUALS) :

        #Selection
        parent1 = selection()
        parent2 = selection()
        while parent1 == parent2:
            parent2 = selection()

        #Crossover
        if random.random() <= CROSSOVER_PROBABILITY:
            offspring = crossover(currentGeneration[parent1], currentGeneration[parent2]);

            #Mutation
            mutate(offspring)

            #Evaluation
            fitnessOffspring = eval(offspring)

            #Recombination 
            if fitnessOffspring <= generationFitness[individual]:
                #delete currentGeneration[individual]
                currentGeneration[individual] = offspring
                generationFitness[individual] = fitnessOffspring

            else:
                pass
                #delete [] offspring;

def selection():
    global generationFitness
    i1 = random.randint(0,N_INDIVIDUALS-1) 
    i2 = random.randint(0,N_INDIVIDUALS-1) 
    while i1 == i2:
        i2 = random.randint(0,N_INDIVIDUALS-1)  

    if generationFitness[i1] < generationFitness[i2]:
        return i1
    else:
        return i2

def getBestSolution():
    global numberOfTests
    global generationFitness
    minFitness = numberOfTests

    for i in range(N_INDIVIDUALS):
        if generationFitness[i] < minFitness:
            minFitness = generationFitness[i]

    return minFitness
def getAverageSolution():
    avgFitness = 0

    for i in range(N_INDIVIDUALS):
        global generationFitness
        avgFitness += generationFitness[i]
        avgFitness /= 1.0*N_INDIVIDUALS

    return avgFitness

def eval(individual):
    fitness = 0
    global numberOfReqs
    satisfiedRequirements = [False for i in range(numberOfReqs)]
    fitness = 0
    for fit in range(numberOfTests):
        fitness = fit
        for req in range(numberOfReqs):
            if reqMatrix[individual[fit]][req] == "1":
                satisfiedRequirements[req] = True
        
        finish = True
        for i in range(numberOfReqs):
            if not satisfiedRequirements[i]:
                finish = False
                break

        if finish:
            break

    return fitness + 1;
def mutate(individual):
    for i in range(numberOfTests):
        if random.random() <= MUTATION_RATE:
            point1 = random.randint(0,numberOfTests-1)
            point2 = random.randint(0,numberOfTests-1)
            while point1 == point2:
                point2 = random.randint(0,numberOfTests-1)

            aux = individual[point1];
            individual[point1] = individual[point2]
            individual[point2] = aux

def crossover(parent1, parent2):
    if len(parent1) != len(set(parent1)):
        print(parent1)
        raise ValueError("parent1 error")
    if len(parent2) != len(set(parent2)):
        print(parent2)
        raise ValueError("parent2 error")
    point1 = random.randint(0, numberOfTests-1)
    point2 = random.randint(point1, numberOfTests-1)

    offspring1=list(parent1)
    add_list=[]
    remove_list=[]

    for i in range(point1, point2+1):
        ind1=parent1.index(parent2[i])
        ind2=parent2.index(parent1[i])
        if (ind1 < point1 or ind1 > point2):
            remove_list.append(parent1[ind1])
        if (ind2 < point1 or ind2 > point2):
            add_list.append(parent2[ind2])
        offspring1[i] = parent2[i]

    for i in range(len(add_list)):
        offspring1[parent1.index(remove_list[i])]=add_list[i]

    offspring2=list(parent2)
    add_list=[]
    remove_list=[]

    for i in range(point1, point2+1):
        ind1=parent2.index(parent1[i])
        ind2=parent1.index(parent2[i])
        if (ind1 < point1 or ind1 > point2):
            remove_list.append(parent2[ind1])
        if (ind2 < point1 or ind2 > point2):
            add_list.append(parent1[ind2])
        offspring2[i] = parent1[i]

    for i in range(len(add_list)):
        offspring2[parent1.index(remove_list[i])]=add_list[i]

    if len(offspring1) != len(set(offspring1)):
        print(offspring1)
        raise ValueError("Offspring1 error")
    if len(offspring2) != len(set(offspring2)):
        print(offspring2)
        raise ValueError("Offspring2 error")

    offFitness1 = eval(offspring1)
    offFitness2 = eval(offspring2)

    if offFitness1 < offFitness2:
        return offspring1
    else:
        return offspring2


start = time.time()
finalGeneration = run()
end = time.time()
print("--- %s seconds ---" % (end - start))
print("\nThe best solution was found in generation "+str(finalGeneration)+", with fitness: "+str(getBestSolution()))
print("Time: "+str(end-start)+" seconds") 

#print(crossover([1,2,3,4,5,6,7,8,9,10],[1,1,1,1,1,1,1,1,1,1]))